File: 2024/src/python/download.py
# Import packages
import os
import requests


def download(day: int, year: int = 2024) -> None:
    """
    Download data from Advent of Code website
    """

    # File where data will be stored
    file = f'data/input{day}.py.txt'

    if os.path.exists(file):
        print(f'Python data for day {day} already downloaded.')
        return

    # URL to download data from
    url = f'https://adventofcode.com/{year}/day/{day}/input'

    # Get session cookie
    with open(f'aoc_{year}_session.txt') as f:
        session_cookie = f.read().strip()

    # Request data
    try:
        response = requests.get(url, cookies={'session': session_cookie})
    except Exception as e:
        print(f'Error downloading Python data for day {day}: {e}')

    # Check if response is 404
    if response.status_code == 404:
        print(f'Python data for day {day} is not available.')
        return

    # Save data to file
    with open(file, 'w') as f:
        f.write(response.text)
    print(f'Python data for day {day} downloaded.')

File: 2024/src/julia/download.jl
module Download

    __precompile__()

    # Import packages
    using HTTP

    function download(day::Int, year::Int = 2024) :: Bool
        """
        Download the input for the given day and year
        """

        # Get the input file
        file :: String = joinpath("data", "input$(day).txt")
        
        # Check if input file already exists
        if isfile(file)
            println("Julia data for day $day already downloaded.")
            return true
        end
        
        # Get the URL
        url = "https://adventofcode.com/$year/day/$day/input"

        # Open session cookie from file
        session_cookie :: String = ""
        open("aoc_$(year)_session.txt") do file
            session_cookie = strip(readline(file))
        end

        # Download response
        try
            response = HTTP.get(url, ["cookie" => "session=$session_cookie"])
            open(file, "w") do file
                write(file, String(response.body))
            end
            println("Julia data for day $day downloaded.")
            return true
        catch e
            if isa(e, HTTP.StatusError) && e.response.status == 404
                println("Julia data for day $day not available.")
                return false
            else
                println("Error downloading Julia data for day $day: $e.")
                return false
            end
        end
    end
end